File: android/src/main/java/top/huic/tencent_im_plugin/message/AbstractMessageNode.java
package top.huic.tencent_im_plugin.message;

import com.tencent.imsdk.v2.V2TIMMessage;

import top.huic.tencent_im_plugin.message.entity.AbstractMessageEntity;

/**
 * 消息节点接口
 *
 * @param <N> 节点类型，对应腾讯云 TIMElem
 */
public abstract class AbstractMessageNode<N, E extends AbstractMessageEntity> {
    /**
     * 获得发送的消息体
     *
     * @param entity 消息实体
     * @return 结果
     */
    public V2TIMMessage getV2TIMMessage(E entity) {
        throw new RuntimeException("This node does not support sending");
    }

    /**
     * 根据消息节点获得描述
     *
     * @param elem 节点
     */
    public abstract String getNote(N elem);

    /**
     * 将节点解析为实体对象
     *
     * @param elem 节点
     * @return 实体对象
     */
    public abstract E analysis(N elem);

    /**
     * 获得实体类型
     *
     * @return 类型
     */
    public abstract Class<E> getEntityClass();
}

File: ios/Classes/message/AbstractMessageNode.swift
import ImSDK

//
//  AbstractMessageNode.swift
//  tencent_im_plugin
//
//  Created by 蒋具宏 on 2020/3/13.
//  消息节点抽象
public class AbstractMessageNode {

    /**
     * 获得发送的消息体
     * @param params 参数
     */
    func getV2TIMMessage(params: [String: Any]) -> V2TIMMessage {
        return V2TIMMessage();
    }

    /**
     * 根据消息节点获得描述
     *
     * @param elem 节点
     */
    func getNote(elem: V2TIMElem) -> String? {
        return nil;
    }

    /// 根据节点解析为实体对象
    func analysis(elem: V2TIMElem) -> AbstractMessageEntity? {
        return nil;
    }

    /// 获得参数
    func getParam<T>(params: [AnyHashable: Any], paramKey: AnyHashable) -> T? {
        if let value = params[paramKey] {
            return (value as! T);
        }
        return nil;
    }
}
